File: src/aa_si_echogram_gl/colormaps.py
"""Colormap export for the web build.

The viewer has to map a value to the same color the existing matplotlib figures
do, so definitions come from matplotlib rather than being written a second time
in TypeScript. This module writes two JSON files: definitions the web build
imports, and a reference the web test asserts against.
"""

import json

LUT_SIZE = 256
"""Entries per colormap. Matches the length of matplotlib's own table, so the
export is the table itself rather than a resampling of it."""

DEFAULT_NAMES = (
    "viridis",
    "magma",
    "inferno",
    "plasma",
    "jet",
    "turbo",
    "gray",
    # Diverging, for difference layers. A difference is signed and symmetric
    # about zero, and a sequential map hides the sign in a brightness ramp.
    "coolwarm",
    "RdBu",
)

NODATA_COLOR = "#2E2E2E"
"""Masked samples, from `cmap.set_bad()` in AA-SI_Visualization."""

NOISE_COLOR = "#000000"
"""Cluster label -1 in AA-SI_Visualization, deliberately receding."""

CLUSTER_PALETTE = (
    "#00F3FC",
    "#35E200",
    "#0400FF",
    "#F943FF",
    "#F30101",
    "#EDFF4D",
    "#4E9200",
    "#970021",
    "#5600C7",
    "#017685",
    "#FFA600",
)
"""Categorical colors for cluster layers, in the order AA-SI_Visualization
assigns them."""


def to_byte(value):
    """Quantize a 0 to 1 channel the way matplotlib does.

    matplotlib truncates rather than rounds, so matching it keeps viewer pixels
    identical to figure pixels instead of off by one.

    Args:
        value: Channel value in the closed interval 0 to 1.

    Returns:
        int: Byte value.
    """
    return min(int(value * 255), 255)


def definition(name, size=LUT_SIZE):
    """Sample a matplotlib colormap as float RGB triples.

    Values are written at full precision. Rounding them could move one across
    a byte boundary, and the point of the export is that the web build lands on
    the same byte matplotlib does.

    Args:
        name: Registered matplotlib colormap name.
        size: Number of entries.

    Returns:
        list: `size` entries, each a list of three floats.
    """
    import numpy as np
    from matplotlib import colormaps

    colors = colormaps[name](np.linspace(0.0, 1.0, size))[:, :3]
    return [[float(c) for c in row] for row in colors]


def reference(name, size=LUT_SIZE):
    """Sample a matplotlib colormap as bytes, using matplotlib's own conversion.

    Args:
        name: Registered matplotlib colormap name.
        size: Number of entries.

    Returns:
        list: Flat RGBA bytes, four per entry.
    """
    import numpy as np
    from matplotlib import colormaps

    rgba = colormaps[name](np.linspace(0.0, 1.0, size), bytes=True)
    return [int(v) for v in rgba.reshape(-1)]


def export(names=None, size=LUT_SIZE):
    """Build the definitions document.

    Args:
        names: Colormap names, or None for DEFAULT_NAMES.
        size: Entries per colormap.

    Returns:
        dict: Serializable definitions.
    """
    names = tuple(names) if names else DEFAULT_NAMES
    return {
        "version": 1,
        "size": size,
        "source": _source(),
        "nodataColor": NODATA_COLOR,
        "noiseColor": NOISE_COLOR,
        "clusterPalette": list(CLUSTER_PALETTE),
        "continuous": {name: definition(name, size) for name in names},
    }


def export_reference(names=None, size=LUT_SIZE):
    """Build the reference document the web test asserts against.

    Args:
        names: Colormap names, or None for DEFAULT_NAMES.
        size: Entries per colormap.

    Returns:
        dict: Serializable reference bytes.
    """
    names = tuple(names) if names else DEFAULT_NAMES
    return {
        "version": 1,
        "size": size,
        "source": _source(),
        "continuous": {name: reference(name, size) for name in names},
    }


def write(path, document):
    """Write a document as JSON.

    Args:
        path: Destination file path.
        document: Result of export or export_reference.
    """
    with open(path, "w", encoding="utf-8") as handle:
        json.dump(document, handle, indent=1)
        handle.write("\n")


def _source():
    """Identify the matplotlib that produced an export."""
    import matplotlib

    return f"matplotlib {matplotlib.__version__}"

File: web/src/render/colormaps.ts
/**
 * Colormap lookup tables.
 *
 * Definitions come from matplotlib through colormaps.json, written by
 * `aa-echogram colormaps`, so a value maps to the same color here as it does in
 * the existing figures. Nothing in this file invents a color.
 */

import definitions from './colormaps.json';

export interface ColormapDocument {
  version: number;
  size: number;
  source: string;
  nodataColor: string;
  noiseColor: string;
  clusterPalette: string[];
  continuous: Record<string, number[][]>;
}

const document = definitions as ColormapDocument;

export const LUT_SIZE = document.size;
export const NODATA_COLOR = parseHex(document.nodataColor);
export const NOISE_COLOR = parseHex(document.noiseColor);

export function colormapNames(): string[] {
  return Object.keys(document.continuous);
}

/**
 * Quantize a 0 to 1 channel the way matplotlib does.
 *
 * matplotlib truncates rather than rounds, and matching it keeps viewer pixels
 * identical to figure pixels rather than off by one.
 */
export function toByte(value: number): number {
  return Math.min(Math.floor(value * 255), 255);
}

/** Build an RGBA byte table for one colormap. */
export function buildLut(name: string): Uint8Array<ArrayBuffer> {
  const stops = document.continuous[name];
  if (!stops) {
    throw new Error(`unknown colormap ${name}; have ${colormapNames().join(', ')}`);
  }
  const lut = new Uint8Array(stops.length * 4);
  for (let i = 0; i < stops.length; i += 1) {
    const [r, g, b] = stops[i];
    lut[i * 4 + 0] = toByte(r);
    lut[i * 4 + 1] = toByte(g);
    lut[i * 4 + 2] = toByte(b);
    lut[i * 4 + 3] = 255;
  }
  return lut;
}

/**
 * Build a table holding one hue at rising alpha.
 *
 * A solid tint is a degenerate colormap rather than a separate shader path, so
 * the tricolor echogram of three tinted layers needs nothing new.
 *
 * The hue is constant and only alpha ramps, which is what makes the additive
 * result linear in the value: blending multiplies rgb by alpha once, so a table
 * that darkened the hue as well would contribute the square of the intensity.
 * Full length rather than two entries, so it samples like any other colormap.
 */
export function buildTintLut(
  tint: [number, number, number],
  size = LUT_SIZE,
): Uint8Array<ArrayBuffer> {
  const lut = new Uint8Array(size * 4);
  for (let i = 0; i < size; i += 1) {
    lut[i * 4 + 0] = toByte(tint[0]);
    lut[i * 4 + 1] = toByte(tint[1]);
    lut[i * 4 + 2] = toByte(tint[2]);
    lut[i * 4 + 3] = toByte(i / (size - 1));
  }
  return lut;
}

/** The categorical base colors, as 0 to 1 triples. */
export function clusterPalette(): [number, number, number][] {
  return document.clusterPalette.map(parseHex);
}

/** Upload a table as a width by 1 rgba8unorm texture. */
export function createLutTexture(
  device: GPUDevice,
  lut: Uint8Array<ArrayBuffer>,
  label = 'colormap',
): GPUTexture {
  const width = lut.length / 4;
  const texture = device.createTexture({
    label,
    size: [width, 1, 1],
    format: 'rgba8unorm',
    usage: GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST,
  });
  device.queue.writeTexture(
    { texture },
    lut,
    { bytesPerRow: width * 4 },
    [width, 1, 1],
  );
  return texture;
}

/** Parse `#rrggbb` into three 0 to 1 channels. */
export function parseHex(hex: string): [number, number, number] {
  const text = hex.replace('#', '');
  if (text.length < 6) throw new Error(`not a hex color: ${hex}`);
  return [
    parseInt(text.slice(0, 2), 16) / 255,
    parseInt(text.slice(2, 4), 16) / 255,
    parseInt(text.slice(4, 6), 16) / 255,
  ];
}
